pkg/server/sandbox_reset_windows.go

```diff
@@ -0,0 +1,256 @@
+//go:build windows
+
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package server
+
+import (
+	"github.com/containerd/containerd"
+	containerdio "github.com/containerd/containerd/cio"
+	"github.com/containerd/containerd/containers"
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/log"
+	"github.com/containerd/containerd/snapshots"
+	customopts "github.com/containerd/cri/pkg/containerd/opts"
+	ctrdutil "github.com/containerd/cri/pkg/containerd/util"
+	"github.com/containerd/cri/pkg/netns"
+	"github.com/containerd/cri/pkg/store"
+	sandboxstore "github.com/containerd/cri/pkg/store/sandbox"
+	"github.com/pkg/errors"
+	"golang.org/x/net/context"
+)
+
+// ResetPodSandbox resets resources and state of the sandbox and its associated containers
+func (c *criService) resetSandbox(ctx context.Context, sandbox sandboxstore.Sandbox) (retErr error) {
+	var (
+		id     = sandbox.ID
+		entity = log.G(ctx).WithField("sandboxID", id)
+	)
+
+	// Return error if sandbox is in running in unknown state
+	state := sandbox.Status.Get().State
+	switch state {
+	case sandboxstore.StateReady:
+		entity.Debugf("sandbox is already running")
+		return nil
+	case sandboxstore.StateNotReady:
+	default:
+		return errors.Errorf("sandbox %q is in invalid state %q", id, state)
+	}
```

Comment:
So I know we don't have a corresponding "Created" state for pods, but does this mean we have no way of telling if a pod has actually run already and exited when we go to reset it?

Reply:
Ish: on pod "creation" (start), the underlying task is automatically started and run, and the pod goes into the ready state.
So, if a pod exists and is not ready, the assumption is that either it was previously running (ready) and was stopped or was loaded (after a containerd/CRI restart) in a stopped or unknown state. From what I can tell, CRI guarantees that pods that fail to start their tasks when created are deleted automatically.